### ISSUE TITLE
Second order gradient for Batch-wise HVP

### ISSUE THREAD

**Issue by haonan3** [f-dangel/backpack#213]
Hi,

Thanks for this great library. I am wondering how can we compute the gradient of sum of gradient?

I am trying to implement the Hessian-Vector-Product (HVP) with the following code:

```
def batch_hvp(self, model, loss, params_list, batch_grad_list):
    if len(params_list) != len(batch_grad_list):
        raise (ValueError("w and v must have the same length."))

    one_sample_grad_list = grad(loss, params_list, retain_graph=True, create_graph=True)

    elemwise_products = 0
    for grad_elem, v_elem in zip(one_sample_grad_list, batch_grad_list):
        sum_over_dims = []
        for i in range(len(v_elem.shape)):
            sum_over_dims.append(i)
        sum_over_dims = tuple(sum_over_dims[1:])
        elemwise_products += torch.sum(grad_elem.unsqueeze(0) * v_elem.detach(), sum_over_dims)

    with backpack(BatchGrad()):
        elemwise_products.backward()   # problem: has no attribute 'input0'
        return_grads = [p.grad_batch for p in model.parameters() if p.requires_grad]

    return return_grads
```
I encounter the  " has no attribute 'input0'  " problem when I call the backward(), is it possible to get batch gradients for return_grads?

For now, I am only using the for loop to compute the gradient.

```
def batch_hvp(self, model, loss, params_list, batch_grad_list):
    if len(params_list) != len(batch_grad_list):
        raise (ValueError("w and v must have the same length."))

    one_sample_grad_list = grad(loss, params_list, retain_graph=True, create_graph=True)

    elemwise_products = 0
    for grad_elem, v_elem in zip(one_sample_grad_list, batch_grad_list):
        sum_over_dims = []
        for i in range(len(v_elem.shape)):
            sum_over_dims.append(i)
        sum_over_dims = tuple(sum_over_dims[1:])
        elemwise_products += torch.sum(grad_elem.unsqueeze(0) * v_elem.detach(), sum_over_dims)
    
    # The for-loop version 
    grad_cache = []
    for i in range(elemwise_products.shape[0]):
        elemwise_products[i].backward(retain_graph=True)
        grad_cache.append([p.grad.clone() for p in model.parameters() if p.requires_grad])
    grad_cache = list(zip(*grad_cache))
    return_grads = []
    for l_id in range(len(grad_cache)):
        return_grads.append(torch.cat([g.unsqueeze(0) for g in grad_cache[l_id]], dim=0))

    return return_grads
```

Thanks in advance!

**Comment by schaefertim** [comment]
Hi,
thanks for providing a detailed description of the problem. I have been able to reproduce the issue.
The issue is that BackPACK does ignore `retain_graph`. In the first backward pass, it deletes its internal variables `input0` and `output`. Those are missing for the second backward pass.
For now, you can try the following workarounds:
1. If the model is deterministic (no BatchNorm, Dropout, ... or in evaluation mode), you can do a second forward pass.
```Python
one_sample_grad_list = grad(loss, params_list, retain_graph=True, create_graph=True)  # first backward
_ = loss_fn(model(inputs), targets)  # second forward to make backpack save its internal variables
```

2. Manually prevent deletion of BackPACK's quantities.
Change https://github.com/f-dangel/backpack/blob/afc94f1d9a1666e56f0cefea24c2288f928f6a56/backpack/__init__.py#L215
to 
```Python
    pass  # memory_cleanup(module)
```
3. Wait until BackPACK supports `retain_graph`.
I already tried to properly implement it. My main issue is, I need to request the value of `retain_graph` from PyTorch from within the backward hook. However, I did not find it. I was expecting something along the lines of `torch.autograd.global_variables.retain_graph`. If you have an idea, please say it.

I hope this clarified things a bit.

**Comment by haonan3** [comment]
Hi,

Thank you very much. I think the second method is better. After I remove `` memory_cleanup``, I just need to manually call  `` memory_cleanup`` by ``model.apply(memory_cleanup)``, for other place where I am using backpack to get grad_batch().

Thanks!

**Comment by f-dangel** [comment]
Hi @haonan3,

did you verify whether your results are in accordance with the Batch-wise HVPs obtained via a for-loop?

**Comment by f-dangel** [comment]
Hi again,

I think your approach for computing IVHPs does not work correctly whenever a parameter is used not only in the forward pass, but also in the gradient computation. [Here](https://github.com/schaefertim/backpack/blob/d1e9e3912890af41f28f3c0eac3e32074bfc3546/docs_src/examples/use_cases/example_indiv_hvp_retain_graph.py) is an example.

Best,
Felix

**Comment by f-dangel** [comment]
Hi again,

just to make sure: Your approach via `BatchGrad` will produce **an incorrect result** (even though the shapes might be correct). Use Hessian-vector products via `autograd` to compute correct IHVPs.

Best,
Felix

**Comment by f-dangel** [comment]
Hey, I did not look at your code in detail, but in principle it should be possible to perform the first backward pass for individual gradients with BackPACK, then use the individual gradients in a `for`-loop to obtain IHVPs via a second backward pass. Please make sure to double-check everything with `autograd`.

Best,
Felix
